mvi-core/src/main/java/ru/coderedwolf/mvi/core/BaseViewModelStore.kt

```diff
@@ -0,0 +1,86 @@
+package ru.coderedwolf.mvi.core
+
+import androidx.annotation.CallSuper
+import androidx.lifecycle.ViewModel
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.rxkotlin.addTo
+import io.reactivex.rxkotlin.withLatestFrom
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.PublishSubject
+import ru.coderedwolf.wordlearn.common.domain.system.SchedulerProvider
+
+/**
+ * @author CodeRedWolf. Date 13.10.2019.
+ */
+abstract class BaseViewModelStore<Action, State, Effect, NavigationEvent>(
+    protected val schedulerProvider: SchedulerProvider,
+    initialState: State,
+    private val reducer: Reducer<State, Effect>,
+    private val middleware: Middleware<Action, State, Effect>,
+    private val navigator: Navigator<State, Effect, NavigationEvent>? = null,
+    private val bootstrapper: Bootstrapper<Action>? = null
+) : ViewModel(), Store<Action, State> {
+
+    private val wiring = CompositeDisposable()
+    private var viewBind: Disposable? = null
+
+    private val stateSubject = BehaviorSubject.createDefault(initialState)
+    private val actionSubject = PublishSubject.create<Action>()
+    private val effectSubject = PublishSubject.create<Effect>()
+    private val navigationEventSubject = PublishSubject.create<NavigationEvent>()
```

Comment:
I decided to rename these vals cause some subjects were named in plural form and some in singular